File: Graph/Python/Find Critical and Pseudo-Critical Edges in Minimum Spanning Tree.py
# Problem link: https://leetcode.com/problems/find-critical-and-pseudo-critical-edges-in-minimum-spanning-tree/description/
# Time: O(e ^ 2 * log(v))
# Space: O(e * v)
class UnionFind:

    def __init__(self, num):
        self.parent = [node for node in range(num)]
        self.size = [1] * num
    
    def find(self, node):
        if self.parent[node] == node:
            return node
        self.parent[node] = self.find(self.parent[node])
        return self.parent[node]
    
    def union(self, nodeOne, nodeTwo):
        parentOne, parentTwo = self.find(nodeOne), self.find(nodeTwo)

        if parentOne != parentTwo:

            if self.size[parentOne] < self.size[parentTwo]:
                self.size[parentTwo] += self.size[parentOne]
                self.parent[parentOne] = parentTwo
            else:
                self.size[parentOne] += self.size[parentTwo]
                self.parent[parentTwo] = parentOne


class Solution:
    def findCriticalAndPseudoCriticalEdges(self, n: int, edges: List[List[int]]) -> List[List[int]]:
        for index, edge in enumerate(edges):
            edge.append(index)          
        
        edges.sort(key = lambda x : x[2])
        unionFind, mstWeight = UnionFind(n), 0

        for ver1, ver2, wei, idx in edges:
            if unionFind.find(ver1) != unionFind.find(ver2):
                unionFind.union(ver1, ver2)
                mstWeight += wei
        
        critical, pseudo = [], []

        for ver1, ver2, wei, idx in edges:
            criticalUF, CWeight = UnionFind(n), 0
            pseudoUF, PCWeight = UnionFind(n), wei
            pseudoUF.union(ver1, ver2)

            for v1, v2, weight, index in edges:
                if index != idx and criticalUF.find(v1) != criticalUF.find(v2):
                    criticalUF.union(v1, v2)
                    CWeight += weight

                if pseudoUF.find(v1) != pseudoUF.find(v2):
                    pseudoUF.union(v1, v2)
                    PCWeight += weight
            
            if CWeight != mstWeight:
                critical.append(idx)
            elif PCWeight == mstWeight:
                pseudo.append(idx)

        return [critical, pseudo]

File: Graph/Java/Find Critical and Pseudo-Critical Edges in Minimum Spanning Tree.java
// Problem link: https://leetcode.com/problems/find-critical-and-pseudo-critical-edges-in-minimum-spanning-tree/description/
// Time: O(e ^ 2 * log(v))
// Space: O(e * v)
class UnionFind {

    int[] parent, size;

    public UnionFind(int num) {
        this.parent = new int[num];
        this.size = new int[num];

        for(int node = 0; node < num; node++) {
            this.parent[node] = node;
            this.size[node] = 1;
        }
    }

    public int find(int node) {
        if(this.parent[node] == node) {
            return node;
        }
        this.parent[node] = this.find(parent[node]);
        return this.parent[node];
    }

    public void union(int nodeOne, int nodeTwo) {
        int parentOne = this.find(nodeOne), parentTwo = this.find(nodeTwo);

        if(parentOne != parentTwo) {
            if(this.size[parentOne] < this.size[parentTwo]) {
                this.size[parentTwo] += this.size[parentOne];
                this.parent[parentOne] = parentTwo;
            }
            else {
                this.size[parentOne] += this.size[parentTwo];
                this.parent[parentTwo] = parentOne;
            }
        }
    }
}


class Solution {
    public List<List<Integer>> findCriticalAndPseudoCriticalEdges(int n, int[][] edges) {
        int size = edges.length, mstWeight = 0, CWeight, PCWeight;
        UnionFind unionFind = new UnionFind(n), CriticalUF, PseudoUF;
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> critical = new ArrayList<>(), pseudo = new ArrayList<>();

        for(int index = 0; index < size; index++) {
            edges[index] = new int[] {edges[index][0], edges[index][1], edges[index][2], index};
        }

        Arrays.sort(edges, (a, b) -> Integer.compare(a[2], b[2]));

        for(int[] edge : edges) {
            if(unionFind.find(edge[0]) != unionFind.find(edge[1])) {
                unionFind.union(edge[0], edge[1]);
                mstWeight += edge[2];
            }
        }

        for(int[] edge : edges) {
            CWeight = 0;
            CriticalUF = new UnionFind(n);

            for(int index = 0; index < size; index++) {
                if(edges[index][3] != edge[3] && CriticalUF.find(edges[index][0]) != CriticalUF.find(edges[index][1])) {
                    CriticalUF.union(edges[index][0], edges[index][1]);
                    CWeight += edges[index][2];
                }
            }

            if(CWeight != mstWeight) {
                critical.add(edge[3]);
            }
            else {
                PCWeight = edge[2];
                PseudoUF = new UnionFind(n);
                PseudoUF.union(edge[0], edge[1]);

                for(int index = 0; index < size; index++) {
                    if(PseudoUF.find(edges[index][0]) != PseudoUF.find(edges[index][1])) {
                        PseudoUF.union(edges[index][0], edges[index][1]);
                        PCWeight += edges[index][2];
                    }
                }

                if(PCWeight == mstWeight) {
                    pseudo.add(edge[3]);
                }
            }
        }

        result.add(critical);
        result.add(pseudo);

        return result;
    }
}
